opencensus/stats/internal/view_data_impl.cc

```diff
@@ -302,5 +304,53 @@ ViewDataImpl::ViewDataImpl(ViewDataImpl* source, absl::Time now)
   }
 }
 
+void ViewDataImpl::SetUpdateTime(const std::vector<std::string>& tag_values,
+                                 absl::Time now) {
+  if (expiry_duration_ == absl::ZeroDuration()) {
+    // No need to track last update time if expiry duration is not set.
+    return;
+  }
+
+  auto update_time_map_iter = update_time_entries_.find(tag_values);
+  if (update_time_map_iter == update_time_entries_.end()) {
+    // The timeseries is not tracked, add it to the update time list and map.
+    update_times_.emplace_front(now, tag_values);
+    update_time_entries_[tag_values] = update_times_.begin();
+  } else {
+    // The timeseries is tracked, update its updated time and move it to the
+    // front of the list.
+    auto update_time_list_iter = update_time_map_iter->second;
+    update_time_list_iter->first = now;
+    update_times_.splice(update_times_.begin(), update_times_,
+                         update_time_list_iter);
+  }
+
+  // Remove data that has not been updated for expiry.
+  while (now - update_times_.back().first > expiry_duration_) {
```

Comment:
For clarity, from here down, it should be called “removeExpired”
You can call it from here. removeExpired must also be called periodically.
We don’t want the data pileup, if the next request comes much later.

Reply:
This function is already triggered periodically, whenever the view has new timeseries create or any update to the existing ones. So the only case that the timeseries got created without being removed is that they are all created during an expiry duration and no update at all after that, which seems to be very rare. Considering this, do you think it is still necessary to call remove periodically?

This seems to be the cleanest way to implement, otherwise we'd either need to plumb RemoveExpired to multiple data structure in order to reuse the existing periodic merging, or create a new timer, which seems to be overkill.

Reply:
Since we are fixing it, we should fix the known cases. I will defer to @g-easy on this.

naming: `SetUpdateTime` name gives no indication that it will purge data.

Reply:
I think purge on update is ok.